### ISSUE TITLE
Base calling Filter for in situ sequencing

### ISSUE THREAD

**Issue by kevinyamauchi** [spacetx/starfish#1263]
# Description
I think it would be useful to create a Filter component for performing base calls. In short, this would compute an image from an `ImageStack` where each channel corresponds to a nucleotide (i.e., C, G, T, A) and each pixel contains the base "quality" The base quality would be a score with range [0, 1] that is the likelihood a particular pixel corresponds to a particular nucleotide.

# Implementation
The filter transforms an image to a nucleotide image by applying a function across the channel dimension to call each pixel for a particular base and returns the confidence score.

In terms of how to score bases, we could adopt something similar to the PHRED ([originally used in capillary-based sequencing](https://en.wikipedia.org/wiki/Phred_base_calling)) or the relative intensity of the channels ([Chen et al. uses the intensity of the strongest channel divided by the root mean square of all four channels for BaristaSeq](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5829746/)). It could be nice to allow users to pass a custom function, but that may not play nice with the logging.

@dganguli , @ambrosejcarr , thoughts?

**Comment by dganguli** [comment]
If I understand correctly, and define space as Z x Y x X, this filter is a RxC x space => RxC x space operation. In the input ImageStack, we have pixel intensity values for each R and C. In the output ImageStack, we have a quality score for each nucleotide (across C) for all rounds R. 

A subsequent decoder could then first perform a transformation of RxC x space => R x space. An example of such a transformation could be an argmax over C.  Finally, each 'called' nucleotide base across R is mapped to the codebook via a lookup table thus creating a R x space => space mapping.

The decoder mentioned above is implemented [here](https://github.com/spacetx/starfish/blob/a35a6233d125e2be7b0e3d8d4be6c201e7e755ae/starfish/core/codebook/codebook.py#L577)  

If I understand this correctly, this Filter may result in better decoding accuracy than simply running the decoder on the raw pixel values themselves. If that's true, I'm all for it! The magic, however, is in specifying the the base scoring function. I really like the idea of allowing users to pass a custom function. This could be done with a lambda from the API which takes a C vector and returns a C vector. We can figure out the logging later.

**Comment by kevinyamauchi** [comment]
I think we are on the same page, @dganguli . I totally agree that choosing the right base quality function is essential for this working. I made a prototype of this in #1281 . It's passing my tests and I ran some simple data through it and I think it works so far. It definitely needs more testing though. The prototype doesn't allow custom base calling quality functions, but if you think the approach makes sense, I think we can add that in.

In terms of decoding downstream, I agree that a "per channel max" type decoder like the one you linked could work well. I am not sure if spot calling on the base call images and then decoding or pixel wise decoding would be best.
